pkg/services/ngalert/schedule/schedule_test.go

```diff
@@ -222,6 +228,78 @@ func TestAlertingTicker(t *testing.T) {
 	})
 }
 
+func TestSendingToExternalAlertmanager(t *testing.T) {
+	dbstore := tests.SetupTestEnv(t, 1)
+	t.Cleanup(registry.ClearOverrides)
+
+	// create alert rule with one second interval
+	alertRule := tests.CreateTestAlertRule(t, dbstore, 1)
+
+	fakeAM := newFakeExternalAlertmanager(t)
+	defer fakeAM.Close()
+
+	// First, let's create an admin configuration that holds an alertmanager.
+	adminConfig := &models.AdminConfiguration{OrgID: 1, Alertmanagers: []string{fakeAM.server.URL}}
+	cmd := store.UpdateAdminConfigurationCmd{AdminConfiguration: adminConfig}
+	require.NoError(t, dbstore.UpdateAdminConfiguration(cmd))
+
+	mockedClock := clock.NewMock()
+	baseInterval := time.Second
+
+	logger := log.New("ngalert schedule test")
+	schedCfg := schedule.SchedulerCfg{
+		C:                mockedClock,
+		BaseInterval:     baseInterval,
+		MaxAttempts:      1,
+		Evaluator:        eval.Evaluator{Cfg: &setting.Cfg{ExpressionsEnabled: true}, Log: logger},
+		RuleStore:        dbstore,
+		InstanceStore:    dbstore,
+		AdminConfigStore: dbstore,
+		Notifier:         &fakeNotifier{},
+		Logger:           logger,
+		Metrics:          metrics.NewMetrics(prometheus.NewRegistry()),
+	}
+	st := state.NewManager(schedCfg.Logger, nilMetrics, dbstore, dbstore)
+	sched := schedule.NewScheduler(schedCfg, nil, "http://localhost", st)
+
+	ctx := context.Background()
+
+	// Make sure we sync the configuration at least once before the evaluation happens to guarantee the sender is running
+	// when the first alert triggers.
+	require.NoError(t, sched.SyncAndApplyConfigFromDatabase())
+
+	// Then, ensure we've discovered the Alertmanager.
+	require.Eventually(t, func() bool {
+		return len(sched.AlertmanagersFor(1)) == 1 && len(sched.DroppedAlertmanagersFor(1)) == 0
+	}, 10*time.Second, 200*time.Millisecond)
```

Comment:
I think we should not explicitly call `SyncAndApplyConfigFromDatabase` since that is not the usual flow. The `go func()` below can come before this and just have `require.Eventually` since in 1s the config should be applied in the background.

Reply:
See my comment above, as a general principle I like unit tests to be deterministic and have setup as I expected them to be - otherwise they become much harder to reason about, flaky given runtime of total poll time my change over time and slower.